### PR TITLE
fixing function app test

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
This fixes and reenables function app test.
Basically test was using fixed name for the resource and that was highly conflict-prone.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
azure_rm_functionapp

##### ANSIBLE VERSION
2.5

### CODE REVIEW

**Comment by zikalino** [comment]
@mattclay please merge this, i updated tests and fixed the issue

File: test/integration/targets/azure_rm_functionapp/tasks/main.yml [hunk 1]
```diff
@@ -1,14 +1,19 @@
+- name: Prepare random number
+  set_fact:
+    rpfx: "{{ resource_group | hash('md5') | truncate(7, True, '') }}{{ 1000 | random }}"
+  run_once: yes
+
 - name: create storage account for function apps
   azure_rm_storageaccount:
     resource_group: '{{ resource_group }}'
-    name: azfunccistor4
+    name: azfunccistor4{{ rpfx }}
```
Comment (mattclay) [inline test/integration/targets/azure_rm_functionapp/tasks/main.yml]: Use the `resource_prefix` variable for a value which is unique to the test run instead of generating one.

This should work:

```yaml
    name: "{{ resource_prefix }}"
```

It is generated by ansible-test on each test run and contains some identifying information about the test as well as a unique value to differentiate it from previous test runs.

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/integration/targets/azure_rm_functionapp/tasks/main.yml (hunk 1 above)
Comment (mattclay) [reply test/integration/targets/azure_rm_functionapp/tasks/main.yml]: If multiple values are needed, use it as a prefix and append an additional value.

Comment (zikalino) [reply test/integration/targets/azure_rm_functionapp/tasks/main.yml]: ok, i will do that. where's resource_prefix coming from? is it something new? i guess i should start updating all my tests....

Comment (mattclay) [reply test/integration/targets/azure_rm_functionapp/tasks/main.yml]: It's generated by `ansible-test` and was part of the initial support for cloud testing:

https://github.com/ansible/ansible/blob/b9b8081a878c60f620deb1d4f6ed4fe1439799de/test/runner/lib/cloud/__init__.py#L362-L374

It is injected as a var when invoking `ansible-playbook` during the test run:

https://github.com/ansible/ansible/blob/b9b8081a878c60f620deb1d4f6ed4fe1439799de/test/runner/lib/cloud/azure.py#L145-L146

**Review by mattclay — APPROVED** [review]

**Comment by zikalino** [comment]
@mattclay 

seems like it can't be used directly as it contains invalid characters. i will try some workaround, but i think it should be fixed in the framework. probably if string contains just alphanumeric characters it should work fine for any name.

fatal: [localhost]: FAILED! => {"changed": false, "msg": "sashippable-58552-62 is not a valid storage account name. Storage account name must be between 3 and 24 characters in length and use numbers and lower-case letters only. - Reason.account_name_invalid"}

**Comment by mattclay** [comment]
For cases where the resource_prefix is too long I recommend calculating the SHA or MD5 of the value and truncating as needed to fit the constraints of the resource being named. If it's just a case of dashes not being allowed, but underscores are, a simple substitution should suffice.

While we could generate a shorter alphanumeric value in the test framework, that reduces the value of the prefix for debugging purposes. For those cases I think it's better the test derives an appropriate value or generates one if necessary.

In other words, use the resource_prefix where it's practical, to ease tracking down resources to the test which originated them.

**Comment by zikalino** [comment]
@mattclay is it ok to merge it now?

File: test/integration/targets/azure_rm_functionapp/tasks/main.yml [hunk 2]
```diff
@@ -1,14 +1,14 @@
 - name: create storage account for function apps
   azure_rm_storageaccount:
     resource_group: '{{ resource_group }}'
-    name: azfunccistor4
+    name: sa{{ resource_prefix | replace("-","x") }}
```
Comment (mattclay) [inline test/integration/targets/azure_rm_functionapp/tasks/main.yml]: Looks good overall. The only thing I'd change is to move the resource_prefix modification to a set_fact task so you can reference a variable without repeating the string replacement on each usage.

**Comment by mattclay** [comment]
@zikalino Looks OK to merge once tests pass.

**Comment by zikalino** [comment]
@mattclay finally this is passing, could you merge?

**Comment by mattclay** [comment]
@zikalino Merged! Thanks for the fixes.
